File: solutions/clojure/cars-assemble/1/src/cars_assemble.clj
(ns cars-assemble)

(defn- get-percentage
  "Retrieve the percentage depending on the speed"
  [speed]
  (cond
    (= 0 speed) 0
    (and (>= speed 1) (<= speed 4)) 100
    (and (>= speed 5) (<= speed 8)) 90
    (= 9 speed) 80
    (= 10 speed) 77
    :else 0
    )
  )

(defn production-rate
  "Returns the assembly line's production rate per hour,
   taking into account its success rate"
  [speed]
  (double (/ (* 221 speed (get-percentage speed)) 100))
  )

(defn working-items
  "Calculates how many working cars are produced per minute"
  [speed]
  (int (/ (production-rate speed) 60))
  )

File: solutions/go/cars-assemble/1/cars_assemble.go
package cars

// CalculateWorkingCarsPerHour calculates how many working cars are
// produced by the assembly line every hour.
func CalculateWorkingCarsPerHour(productionRate int, successRate float64) float64 {
	return float64(productionRate) / 100 * successRate
}

// CalculateWorkingCarsPerMinute calculates how many working cars are
// produced by the assembly line every minute.
func CalculateWorkingCarsPerMinute(productionRate int, successRate float64) int {
	perHour := CalculateWorkingCarsPerHour(productionRate, successRate)
	return int(perHour) / 60
}

// CalculateCost works out the cost of producing the given number of cars.
func CalculateCost(carsCount int) uint {
	rest := carsCount % 10
	discounted_tens := (carsCount - rest) / 10
	return uint(rest*10_000 + discounted_tens*95_000)
}
